File: frontend/src/utils/validators.ts
/**
 * Validation utilities for Indonesian documents and identifiers
 */

// ============= NPWP Validation =============

/**
 * Validate Individual NPWP (16-digit NIK)
 * As of July 1, 2024: Residents use 16-digit NIK as NPWP
 * Format: 16 digits with no separators
 */
export const validateIndividualNPWP = (npwp: string): boolean => {
  const regex = /^\d{16}$/;
  return regex.test(npwp.replace(/\D/g, ''));
};

/**
 * Validate Company NPWP (16-digit with leading zero)
 * Format: 0 followed by 15 digits
 */
export const validateCompanyNPWP = (npwp: string): boolean => {
  const regex = /^0\d{15}$/;
  return regex.test(npwp.replace(/\D/g, ''));
};

/**
 * Format NPWP for display
 * 16 digits → XX.XXX.XXX.X-XXX.XXX
 */
export const formatNPWP = (npwp: string): string => {
  const clean = npwp.replace(/\D/g, '');
  if (clean.length !== 16) return npwp;
  return `${clean.slice(0, 2)}.${clean.slice(2, 5)}.${clean.slice(5, 8)}.${clean.slice(8, 9)}-${clean.slice(9, 12)}.${clean.slice(12, 16)}`;
};

/**
 * Parse formatted NPWP back to raw numbers
 */
export const parseNPWP = (formatted: string): string => {
  return formatted.replace(/\D/g, '');
};

// ============= NIK Validation =============

/**
 * Validate Indonesian NIK (16-digit National Identity Number)
 * Format: 16 consecutive digits
 * Structure: DDMMYY-PPPP-CCSX
 *   DD: Day (01-31)
 *   MM: Month (01-12, modify for female: 40-51)
 *   YY: Year (00-99)
 *   PPPP: Province code
 *   CC: City code
 *   S: Series number
 *   X: Registration number
 */
export const validateNIK = (nik: string): boolean => {
  const regex = /^\d{16}$/;
  const clean = nik.replace(/\D/g, '');
  
  if (!regex.test(clean)) return false;
  
  // Validate day: 01-31 or 41-71 (female modifier)
  const day = parseInt(clean.slice(0, 2));
  if ((day < 1 || day > 31) && (day < 41 || day > 71)) return false;
  
  // Validate month: 01-12
  const month = parseInt(clean.slice(2, 4));
  if (month < 1 || month > 12) return false;
  
  return true;
};

/**
 * Format NIK for display
 * 16 digits → XXXX.XXXX.XXXXXXX
 */
export const formatNIK = (nik: string): string => {
  const clean = nik.replace(/\D/g, '');
  if (clean.length !== 16) return nik;
  return `${clean.slice(0, 4)}.${clean.slice(4, 8)}.${clean.slice(8, 16)}`;
};

// ============= Phone Validation =============

/**
 * Validate Indonesian phone number
 * Formats: 08XX-XXXX-XXXX or +62 8XX-XXXX-XXXX
 */
export const validatePhone = (phone: string): boolean => {
  const regex = /^(\+62|0)[0-9]{9,12}$/;
  return regex.test(phone.replace(/\D/g, '+').replace('+', ''));
};

/**
 * Format phone number
 * 0812345678 → 0812-3456-78
 */
export const formatPhone = (phone: string): string => {
  const clean = phone.replace(/\D/g, '');
  if (clean.length < 10) return phone;
  
  if (clean.startsWith('62')) {
    const formatted = clean.slice(2);
    return `+62 ${formatted.slice(0, 3)}-${formatted.slice(3, 7)}-${formatted.slice(7)}`;
  } else if (clean.startsWith('0')) {
    return `${clean.slice(0, 4)}-${clean.slice(4, 8)}-${clean.slice(8)}`;
  }
  return phone;
};

// ============= Currency Validation =============

/**
 * Validate Indonesian Rupiah amount (positive integer)
 */
export const validateCurrency = (value: number): boolean => {
  return !isNaN(value) && value >= 0 && Number.isInteger(value);
};

/**
 * Format number as Indonesian currency
 * 5000000 → Rp 5.000.000
 */
export const formatCurrency = (value: number): string => {
  if (!validateCurrency(value)) return 'Rp 0';
  return new Intl.NumberFormat('id-ID', {
    style: 'currency',
    currency: 'IDR',
    minimumFractionDigits: 0,
    maximumFractionDigits: 0,
  }).format(value);
};

/**
 * Parse currency string to number
 * "Rp 5.000.000" → 5000000
 */
export const parseCurrency = (formatted: string): number => {
  const cleaned = formatted.replace(/\D/g, '');
  return parseInt(cleaned, 10) || 0;
};

// ============= Email Validation =============

/**
 * Validate email address
 */
export const validateEmail = (email: string): boolean => {
  const regex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
  return regex.test(email);
};

// ============= Date Validation =============

/**
 * Validate date is in valid range and format
 */
export const validateDate = (date: string): boolean => {
  try {
    const d = new Date(date);
    return d instanceof Date && !isNaN(d.getTime());
  } catch {
    return false;
  }
};

// ============= Composite Validation =============

export const validateCompanyData = (company: any): { valid: boolean; errors: string[] } => {
  const errors: string[] = [];
  
  if (!company.nama || company.nama.trim().length === 0) {
    errors.push('Nama perusahaan wajib diisi');
  }
  
  if (!validateCompanyNPWP(company.npwp)) {
    errors.push('NPWP perusahaan tidak valid (format: 0XXXXXXXXXXXXXXX)');
  }
  
  if (!company.alamat || company.alamat.trim().length === 0) {
    errors.push('Alamat wajib diisi');
  }
  
  if (company.email && !validateEmail(company.email)) {
    errors.push('Format email tidak valid');
  }
  
  if (company.telepon && !validatePhone(company.telepon)) {
    errors.push('Format telepon tidak valid');
  }
  
  return {
    valid: errors.length === 0,
    errors,
  };
};

export const validateEmployeeData = (employee: any): { valid: boolean; errors: string[] } => {
  const errors: string[] = [];
  
  if (!employee.nama || employee.nama.trim().length === 0) {
    errors.push('Nama karyawan wajib diisi');
  }
  
  if (!validateIndividualNPWP(employee.npwp)) {
    errors.push('NPWP karyawan tidak valid (16 digit)');
  }
  
  if (!validateNIK(employee.nik)) {
    errors.push('NIK tidak valid (16 digit)');
  }
  
  if (!employee.tgl_lahir || !validateDate(employee.tgl_lahir)) {
    errors.push('Tanggal lahir tidak valid');
  }
  
  if (!employee.jenis_kelamin || !['L', 'P'].includes(employee.jenis_kelamin)) {
    errors.push('Jenis kelamin harus L atau P');
  }
  
  if (!employee.jabatan || employee.jabatan.trim().length === 0) {
    errors.push('Jabatan wajib diisi');
  }
  
  if (!employee.status_ptkp || employee.status_ptkp.trim().length === 0) {
    errors.push('Status PTKP wajib diisi');
  }
  
  if (!employee.salary_profile || !validateCurrency(employee.salary_profile.gaji_pokok)) {
    errors.push('Gaji pokok harus berupa angka positif');
  }
  
  return {
    valid: errors.length === 0,
    errors,
  };
};

File: app/core/validators.py
"""
Backend validation utilities
Indonesian Payroll System - Production Grade
"""

import re
from typing import Tuple, Optional
from datetime import datetime


def validate_company_npwp(npwp: Optional[str]) -> bool:
    """
    Validate company NPWP format
    Format: 0XXXXXXXXXXXXXXX (starts with 0, followed by 15 digits)
    """
    if not npwp:
        return False
    
    # Remove whitespace
    npwp = npwp.strip()
    
    # Check length
    if len(npwp) != 16:
        return False
    
    # Check format: starts with 0, all digits
    pattern = r"^0\d{15}$"
    return bool(re.match(pattern, npwp))


def validate_individual_npwp(npwp: Optional[str]) -> bool:
    """
    Validate individual NPWP format
    Format: XXXXXXXXXXXXXXXX (16 digits)
    Based on NIK (national ID)
    """
    if not npwp:
        return False
    
    # Remove whitespace
    npwp = npwp.strip()
    
    # Check length
    if len(npwp) != 16:
        return False
    
    # Check all digits
    pattern = r"^\d{16}$"
    return bool(re.match(pattern, npwp))


def validate_nik(nik: Optional[str]) -> bool:
    """
    Validate Indonesian NIK (National ID)
    Format: XXXXXXXXXXXXXXXX (16 digits)
    
    Structure:
    - 6 digits: birth date (YYMMDD)
    - 5 digits: birth location code
    - 4 digits: serial number
    - 1 digit: gender (0-4 = female, 5-9 = male)
    """
    if not nik:
        return False
    
    # Remove whitespace
    nik = nik.strip()
    
    # Check length
    if len(nik) != 16:
        return False
    
    # Check all digits
    if not nik.isdigit():
        return False
    
    # Validate date portion (YYMMDD)
    birth_date_str = nik[:6]
    
    try:
        # Parse with century prefix (19 or 20)
        if int(birth_date_str[:2]) > 50:
            full_date = f"19{birth_date_str}"
        else:
            full_date = f"20{birth_date_str}"
        
        # Validate it's a valid date
        datetime.strptime(full_date, "%Y%m%d")
    except (ValueError, IndexError):
        return False
    
    # Gender digit validation (last digit before end)
    gender_digit = int(nik[8])
    if gender_digit > 9:  # Should be 0-9
        return False
    
    return True


def validate_phone_number(phone: Optional[str]) -> bool:
    """
    Validate Indonesian phone number
    Formats:
    - 08XX XXXX XXXX (with spaces)
    - 08XXXXXXXXXX (no spaces)
    - +628XX XXXX XXXX (international)
    - 628XXXXXXXXXX (international without +)
    """
    if not phone:
        return False
    
    # Remove common formatting characters
    phone = re.sub(r'[\s\-().]', '', phone)
    
    # Should be 10-13 digits, start with 8 or 62
    if re.match(r"^8\d{9,11}$", phone):
        return True
    
    if re.match(r"^628\d{8,10}$", phone):
        return True
    
    return False


def validate_email(email: Optional[str]) -> bool:
    """
    Validate email format
    """
    if not email:
        return False
    
    pattern = r"^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$"
    return bool(re.match(pattern, email))


def validate_currency(value: Optional[float]) -> bool:
    """
    Validate currency amount
    Must be positive integer
    """
    if value is None:
        return False
    
    try:
        # Convert to int to check if whole number
        val_int = int(value)
        return val_int >= 0
    except (ValueError, TypeError):
        return False


def validate_ptkp_status(status: Optional[str]) -> bool:
    """
    Validate PTKP (personal tax allowance) status
    Valid statuses: TK/0, K/0, K/1, K/2, K/3, etc.
    """
    if not status:
        return False
    
    # Valid patterns: TK/0, K/X where X is digit
    pattern = r"^(TK|K)/\d+$"
    return bool(re.match(pattern, status.upper()))


def validate_company_data(company_data: dict) -> Tuple[bool, list[str]]:
    """
    Validate complete company data
    Returns: (is_valid, error_messages)
    """
    errors = []
    
    # Validate nama (company name)
    if not company_data.get("nama") or len(company_data.get("nama", "")) < 2:
        errors.append("Nama perusahaan harus minimal 2 karakter")
    
    # Validate NPWP
    if not company_data.get("npwp"):
        errors.append("NPWP perusahaan harus diisi")
    elif not validate_company_npwp(company_data.get("npwp")):
        errors.append("Format NPWP perusahaan tidak valid (harus 0XXXXXXXXXXXXXXX)")
    
    # Validate address
    if not company_data.get("alamat") or len(company_data.get("alamat", "")) < 5:
        errors.append("Alamat harus minimal 5 karakter")
    
    # Validate city
    if not company_data.get("kota"):
        errors.append("Kota harus diisi")
    
    # Validate province
    if not company_data.get("provinsi"):
        errors.append("Provinsi harus diisi")
    
    # Validate email if provided
    if company_data.get("email") and not validate_email(company_data.get("email")):
        errors.append("Format email tidak valid")
    
    # Validate phone if provided
    if company_data.get("telepon") and not validate_phone_number(company_data.get("telepon")):
        errors.append("Format nomor telepon tidak valid")
    
    return len(errors) == 0, errors


def validate_employee_data(employee_data: dict) -> Tuple[bool, list[str]]:
    """
    Validate complete employee data
    Returns: (is_valid, error_messages)
    """
    errors = []
    
    # Validate name
    if not employee_data.get("nama") or len(employee_data.get("nama", "")) < 2:
        errors.append("Nama karyawan harus minimal 2 karakter")
    
    # Validate NIK
    if not employee_data.get("nik"):
        errors.append("NIK harus diisi")
    elif not validate_nik(employee_data.get("nik")):
        errors.append("Format NIK tidak valid (harus 16 digit)")
    
    # Validate NPWP if provided
    if employee_data.get("npwp") and not validate_individual_npwp(employee_data.get("npwp")):
        errors.append("Format NPWP tidak valid (harus 16 digit)")
    
    # Validate PTKP status
    if not employee_data.get("status_ptkp"):
        errors.append("Status PTKP harus diisi")
    elif not validate_ptkp_status(employee_data.get("status_ptkp")):
        errors.append("Status PTKP tidak valid (contoh: TK/0, K/0, K/1, etc.)")
    
    return len(errors) == 0, errors


__all__ = [
    "validate_company_npwp",
    "validate_individual_npwp",
    "validate_nik",
    "validate_phone_number",
    "validate_email",
    "validate_currency",
    "validate_ptkp_status",
    "validate_company_data",
    "validate_employee_data",
]
